lib/zhuyin.ex

```diff
@@ -0,0 +1,326 @@
+    "ㄓ" => "zh",
+    "ㄔ" => "ch",
+    "ㄕ" => "sh",
+    "ㄖ" => "r",
+    "ㄗ" => "z",
+    "ㄘ" => "c",
+    "ㄙ" => "s",
+  }
+
+  @standalone_finals %{
+    "ㄧ" => "yi",
+    "ㄨ" => "wu",
+    "ㄩ" => "yu",
+    "ㄧㄚ" => "ya",
+    "ㄨㄚ" => "wa",
+    "ㄧㄥ" => "ying",
+    "ㄧㄤ" => "yang",
+    "ㄧㄝ" => "ye",
+    "ㄨㄛ" => "wo",
+    "ㄨㄥ" => "weng",
+    "ㄨㄤ" => "wang",
+    "ㄧㄠ" => "yaog",
+    "ㄨㄞ" => "waig",
+    "ㄩㄝ" => "yue",
+    "ㄩㄥ" => "yong",
+    "ㄧㄡ" => "you",
+    "ㄨㄟ" => "wei",
+    "ㄧㄢ" => "yan",
+    "ㄨㄢ" => "wan",
+    "ㄩㄢ" => "yuan",
+    "ㄧㄣ" => "yin",
+    "ㄨㄣ" => "wen",
+    "ㄩㄣ" => "yun",
+    # Standalone initials. But parsed as finals for convenience
+    "ㄓ" => "zhi",
+    "ㄔ" => "chi",
+    "ㄕ" => "shi",
+    "ㄖ" => "ri",
+    "ㄗ" => "zi",
+    "ㄘ" => "ci",
+    "ㄙ" => "si",
+    # Standalone finals that are the same as combined
+    "ㄦ" => "er",
+    "ㄢ" => "an",
+  }
+  @finals %{
+    "ㄧ" => "i",
+    "ㄨ" => "u",
+    "ㄩ" => "v",
+    "ㄚ" => "a",
+    "ㄛ" => "o",
+    "ㄜ" => "e",
+    "ㄝ" => "e",
+    "ㄞ" => "ai",
+    "ㄟ" => "ei",
+    "ㄠ" => "ao",
+    "ㄡ" => "ou",
+    "ㄢ" => "an",
+    "ㄣ" => "en",
+    "ㄤ" => "ang",
+    "ㄥ" => "eng",
+    "ㄦ" => "er",
+    "ㄧㄚ" => "ia",
+    "ㄨㄚ" => "ua",
+    "ㄧㄥ" => "ing",
+    "ㄧㄤ" => "iang",
+    "ㄧㄝ" => "ie",
+    "ㄨㄛ" => "uo",
+    "ㄨㄥ" => "ong",
+    "ㄨㄤ" => "uang",
+    "ㄧㄠ" => "iao",
+    "ㄨㄞ" => "uai",
+    "ㄩㄝ" => "ve",
+    "ㄩㄥ" => "iong",
+    "ㄧㄡ" => "iu",
+    "ㄨㄟ" => "ui",
+    "ㄧㄢ" => "ian",
+    "ㄨㄢ" => "uan",
+    "ㄩㄢ" => "van",
+    "ㄧㄣ" => "in",
+    "ㄨㄣ" => "un",
+    "ㄩㄣ" => "vn",
+  }
+
+  @doc"""
+  Convert zhuyin to pinyin
+
+  # Examples
+   iex>  decode_zhuyin("ㄝ")
+   None
+
+   iex>  decode_zhuyin("ㄐㄩ˙")
+   "jv"
+
+   iex>  decode_zhuyin("ㄌㄩˇ")
+   "lv3"
+
+   iex>  decode_zhuyin("ㄓㄨㄢˋ")
+   "zhuan4"
+
+   iex>  decode_zhuyin("ㄓㄠˊ")
+   "zhao2"
+
+   iex>  decode_zhuyin("ㄓˋ")
+   "zhi4"
+
+   iex>  decode_zhuyin("ㄌㄥ")
+   "leng1"
+
+   iex>  decode_zhuyin("ㄕㄨㄟˇ")
+   "shui3"
+
+   iex>  decode_zhuyin("ㄌㄧㄡˊ")
+   "liu2"
+
+   iex>  decode_zhuyin("ㄧˊ")
+   "yi2"
+
+   iex>  decode_zhuyin("ㄇㄧㄣˊ")
+   "min2"
+
+   iex>  decode_zhuyin("ㄐˇ")
+   None
+
+   // Zhuyin specific
+   iex>  decode_zhuyin("ㄨㄥˊ")
+   "weng2"
+
+   iex>  decode_zhuyin("ㄨˊ")
+   "wu2"
+
+   iex>  decode_zhuyin("ㄩ")
+   "yu1"
+
+   iex>  decode_zhuyin("ㄩㄝˇ")
+   "yue3"
+
+   iex>  decode_zhuyin("ㄩㄝˋ")
+   "yue4"
+
+   iex>  decode_zhuyin("ㄩㄢˊ")
+   "yuan2"
+
+   iex>  decode_zhuyin("ㄌㄩㄢˋ")
+   "lvan4"
+
+   iex>  decode_zhuyin("ㄌㄢˇ")
+   "lan3"
+
+   iex>  decode_zhuyin("ㄦˊ")
+   "er2"
+
+   iex>  decode_zhuyin("a5")
+   None
+
+   // Special case
+   iex>  decode_zhuyin("ㄦ˙")
+   "r"
+  """
+  # @spec decode_zhuyin(String.t()) :: Integer.t()
+   def decode_zhuyin(input) do
+    case input |> Pinyin.ZhuYinParsers.zhuyin_word |> IO.inspect do
+      {:ok, parsed, _rest, _other1, _other2, _other3} ->
+        first = parsed |> List.first
+        if first == input do
+          None
```

Comment:
Should use `nil` here, right?

Reply:
Or maybe `{:error, ...}`

Reply:
`None` is a module name, so `nil` is what you are looking for.

However, in general, `nil` not used to signal errors. Generally the function returns something  like`{:ok, result}` or `{:error, reason}`, if errors are a possibility. A `function_name!` version is then introduced which raises an exception if an error occurs and just returns `result` otherwise.